### ISSUE TITLE
Port all Scala code to Java

### ISSUE THREAD

**Issue by Scott-Guest** [runtimeverification/k#4030]
While Scala is a nice language, the consensus of the frontend team is that maintaining a mixture of Scala and Java causes more friction than it is worth:
- IntelliJ struggles to reason about the Scala code and reports false errors (#3867)
- We have to frequently convert between Java and Scala types (see the usages of [`collections.scala`](https://github.com/runtimeverification/k/blob/08477b587642c382e3ff11d7983431bfa3e8c124/kore/src/main/scala/org/kframework/collections.scala#L17)) and write boilerplate code to offer two versions of the same methods (e.g. `getOption` and `getOptional` in `Att.scala`).
- As a company, my impression is that more of us are familiar with the intricacies of Java than of Scala.

As a result, we have decided that
- All new code should be written in Java
- In the longterm, all existing Scala code should be ported to Java as well

The largest blocker will be the [`matching`](https://github.com/runtimeverification/llvm-backend/tree/34a33b014fbd41fce1059feff37f462c82d5a4c5/matching) library in the LLVM backend, which makes heavy usage of pattern matching and likely can't be ported until we update to Java 21 (#3499).

However, [searching the frontend](https://github.com/search?q=repo%3Aruntimeverification%2Fk+path%3A*.scala&type=Code), there are only 47 Scala files, and many of them can be straightforwardly ported even without pattern matching support in Java.

A high level roadmap:
- [ ] Port Scala case classes not involved in pattern matching to Java record classes inheriting from a sealed interface
    - [`Rich.scala`](https://github.com/runtimeverification/k/blob/08477b587642c382e3ff11d7983431bfa3e8c124/kore/src/main/scala/org/kframework/kore/Rich.scala), [`LabelInfoFromModule.scala`](https://github.com/runtimeverification/k/blob/08477b587642c382e3ff11d7983431bfa3e8c124/kore/src/main/scala/org/kframework/compile/LabelInfoFromModule.scala), [`Location.scala`](https://github.com/runtimeverification/k/blob/08477b587642c382e3ff11d7983431bfa3e8c124/kore/src/main/scala/org/kframework/attributes/Location.scala), etc.
    - The main KORE data structures likely can't be ported due to their use in `matching`
- [ ] Port Scala classes/algorithms not making use of pattern matching
   - [`POSet.scala`](https://github.com/runtimeverification/k/blob/08477b587642c382e3ff11d7983431bfa3e8c124/kore/src/main/scala/org/kframework/POSet.scala), [`Att.scala`](https://github.com/runtimeverification/k/blob/08477b587642c382e3ff11d7983431bfa3e8c124/kore/src/main/scala/org/kframework/attributes/Att.scala), etc.
- [ ] Make remaining Scala classes as Java-esque as possible
   - [ ] Port any `trait` to an `interface`
   - [ ] Use Java types for all members that are not pattern matched on
       - We already have `pcollections` as a dependency, which provides immutable functional data structures
 - [ ] Update to Java 21
 - [ ] Port remaining Scala code

**Comment by Scott-Guest** [comment]
Another option might be to extend @Baltoli's [pattern matching library](https://github.com/runtimeverification/llvm-backend/blob/34a33b014fbd41fce1059feff37f462c82d5a4c5/include/kllvm/ast/pattern_matching.h) then port `matching` to C++.

This would be more work in the short term, but would also make the LLVM backend more cohesive and get rid of the reverse dependency on K.

**Comment by ehildenb** [comment]
Love the idea of removing the reverse dependency! LLVM backend being self-contained seems worth the effort. That matching code is just used for LLVM codegen right?

**Comment by Baltoli** [comment]
Yep, that's the only place it's used. I think we'd want to first build a C++ implementation that generates the same YAML as the scala code does, test equivalence of the outputs, then eliminate the intermediate YAML. It would be a reasonably big rewrite but Dwight's documentation is really good, and we have a strong way of testing any new implementation!
